### PR TITLE
server: add global configuration for default router service offering

### PR DESCRIPTION
**Pull request by ustcweizhou**
## Description

This adds a global setting router.service.offering, which can be uuid of a system offering. 
the service offering will be used by virtual routers; 
if it is not set, default system offering will be used.

This is a account-scope setting, so it can be set in Account setting as well.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?

1. create a system offering
2. change global setting 
image

3. restart network with cleanup
4. new VRs will be created with new offering
image

### CODE REVIEW

**Comment by svenvogel** [comment]
@ustcweizhou nice feature.

**Comment by harikrishna-patnala** [comment]
We already have a provision to mention the service offering id while creating network offering which serves the same purpose right!

**Comment by weizhouapache** [comment]
> We already have a provision to mention the service offering id while creating network offering which serves the same purpose right!

@harikrishna-patnala it is an option, but it is complicated to change the offering of a VR :
(1)create network offering with router offering id
(2) update network to new network offering
Suppose there are many networks with different network offerings, we need to create some new offerings. There is some downtime in step (2).

It is very simple with this PR. (1) change setting, (2) restart network with cleanup.

**Comment by harikrishna-patnala** [comment]
I'm not against the new option, but wanted to know the main intention of adding this. 
Code changes LGTM.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
I need a test volunteer, other than
@blueorangutan test

**Comment by harikrishna-patnala** [comment]
> I need a test volunteer, other than
> @blueorangutan test

I'll test this and post the results

**Comment by weizhouapache** [comment]
> > I need a test volunteer, other than
> > @blueorangutan test
> 
> I'll test this and post the results

@harikrishna-patnala gooood. if you need my assistance, let me know.

**Comment by harikrishna-patnala** [comment]
Following are the test cases covered for both isolated and VPC networks.

1. Created a network with '**router.service.offering**' set to **empty** - VR deployed with '**System Offering For Software Router**' - PASSED
2.  Created a network with **global** configuration parameter **'router.service.offering**' set to UUID of '**VRcustomCO1**' compute offering - VR deployed with '**VRcustomCO1**' - PASSED
3. Created a network with **account level** configuration parameter '**router.service.offering**' set to UUID of '**VRcustomCO2**' compute offering - VR deployed with '**VRcustomCO2**' - PASSED
4. Created a network with **account level** configuration parameter '**router.service.offering**' set to an **invalid UUID** - VR deployed with '**System Offering For Software Router**'. Here in this case global configuration parameter is set to '**VRcustomCO1**'.

In test case 4, irrespective of global config parameter VR deployed with default SO. If this behavior is expected a log message would help debugging.

**Comment by weizhouapache** [comment]
> Following are the test cases covered for both isolated and VPC networks.
> 
> 1. Created a network with '**router.service.offering**' set to **empty** - VR deployed with '**System Offering For Software Router**' - PASSED
> 2. Created a network with **global** configuration parameter **'router.service.offering**' set to UUID of '**VRcustomCO1**' compute offering - VR deployed with '**VRcustomCO1**' - PASSED
> 3. Created a network with **account level** configuration parameter '**router.service.offering**' set to UUID of '**VRcustomCO2**' compute offering - VR deployed with '**VRcustomCO2**' - PASSED
> 4. Created a network with **account level** configuration parameter '**router.service.offering**' set to an **invalid UUID** - VR deployed with '**System Offering For Software Router**'. Here in this case global configuration parameter is set to '**VRcustomCO1**'.
> 
> In test case 4, irrespective of global config parameter VR deployed with default SO. If this behavior is expected a log message would help debugging.

@harikrishna-patnala thanks for testing. I think it make more sense to use '**VRcustomCO1**' instead of '**System Offering For Software Router**' in case 4.  I will update the PR.

File: server/src/main/java/org/cloud/network/router/deployment/RouterDeploymentDefinition.java [hunk 1]
```diff
@@ -389,8 +390,34 @@ protected void findDefaultServiceOfferingId() {
...
+        }
+    }
+
+    private void verifyServiceOfferingByUuid(String offeringUuid) {
+        logger.debug("Verifying router service offering with uuid : " + offeringUuid);
+        ServiceOfferingVO serviceOffering = serviceOfferingDao.findByUuid(offeringUuid);
+        if (serviceOffering != null && serviceOffering.isSystemUse()) {
+            boolean isLocalStorage = ConfigurationManagerImpl.SystemVMUseLocalStorage.valueIn(dest.getDataCenter().getId());
+            if (isLocalStorage == serviceOffering.isUseLocalStorage()) {
+                logger.debug("service offering " + serviceOffering.getId() + " will be used on virtual router");
```
Comment (harikrishna-patnala) [inline server/src/main/java/org/cloud/network/router/deployment/RouterDeploymentDefinition.java]: offeringUuid would be better than id in the log message. with that LGTM

**Comment by weizhouapache** [comment]
updated based on @harikrishna-patnala 's comment

File: server/src/main/java/org/cloud/network/router/deployment/RouterDeploymentDefinition.java [hunk 2]
```diff
@@ -389,8 +390,34 @@ protected void findDefaultServiceOfferingId() {
         serviceOfferingId = serviceOffering.getId();
     }
 
+    protected void findAccountServiceOfferingId(long accountId) {
+        String accountRouterOffering = VirtualNetworkApplianceManager.VirtualRouterServiceOffering.valueIn(accountId);
+        String globalRouterOffering = VirtualNetworkApplianceManager.VirtualRouterServiceOffering.value();
+        if (accountRouterOffering != null) {
+            verifyServiceOfferingByUuid(accountRouterOffering);
+        }
+        if (serviceOfferingId == null && globalRouterOffering != accountRouterOffering) {
```
Comment (DaanHoogland) [inline server/src/main/java/org/cloud/network/router/deployment/RouterDeploymentDefinition.java]: Wouldn't we want `globalRouterOffering` to take precedence over `serviceOfferingId`, @weizhouapache?

Comment (weizhouapache) [reply server/src/main/java/org/cloud/network/router/deployment/RouterDeploymentDefinition.java]: @DaanHoogland serviceOfferingId is the final offering which will be used on virtual router.

Comment (DaanHoogland) [reply server/src/main/java/org/cloud/network/router/deployment/RouterDeploymentDefinition.java]: yes, but if we configured a global router offering wouldn't we prefer that even if serviceOfferingId is not null?

Comment (weizhouapache) [reply server/src/main/java/org/cloud/network/router/deployment/RouterDeploymentDefinition.java]: @daanhoogland correct. it checks following offerings

1. Router service Offering of network/vpc offering
2. Account route service offering
3. Global route service offering 
4. Default route offering 

If 1 or 2 is set and validated, global setting will not be used.

Comment (DaanHoogland) [reply server/src/main/java/org/cloud/network/router/deployment/RouterDeploymentDefinition.java]: @weizhouapache `if (serviceOffering != null)`, which should be always, `globalRouterOffering` will not be checked, whether or not `globalRouterOffering` == `accountRouterOffering`.
so I don't think your 3 and 4 are correct but this condition will never be met and `verifyServiceOfferingByUuid(globalRouterOffering)` will never be executed.
So, why check for `serviceOfferingId` before `globalRouterOffering` when it is the last resort and is only relevant after `globalRouterOffering`?

Comment (weizhouapache) [reply server/src/main/java/org/cloud/network/router/deployment/RouterDeploymentDefinition.java]: > @weizhouapache `if (serviceOffering != null)`, which should be always, `globalRouterOffering` will not be checked, whether or not `globalRouterOffering` == `accountRouterOffering`.
> so I don't think your 3 and 4 are correct but this condition will never be met and `verifyServiceOfferingByUuid(globalRouterOffering)` will never be executed.
> So, why check for `serviceOfferingId` before `globalRouterOffering` when it is the last resort and is only relevant after `globalRouterOffering`?

@DaanHoogland it is serviceOfferingId , not serviceOffering.

https://github.com/apache/cloudstack/blob/3783a092a9431e64f6cf11fc7a1056b338f4f7ab/server/src/main/java/org/cloud/network/router/deployment/RouterDeploymentDefinition.java#L109


the check procedure

https://github.com/apache/cloudstack/blob/3783a092a9431e64f6cf11fc7a1056b338f4f7ab/server/src/main/java/org/cloud/network/router/deployment/RouterDeploymentDefinition.java#L416-L424


in the account part, considering the uuid might be invalid, so check account setting at first, then global setting if account setting is invalid. (if account setting is not set, accountRouterOffering is actually the global setting, so skip the check if accountRouterOffering equals to globalRouterOffering )

https://github.com/apache/cloudstack/blob/3783a092a9431e64f6cf11fc7a1056b338f4f7ab/server/src/main/java/org/cloud/network/router/deployment/RouterDeploymentDefinition.java#L393-L414


hope it helps you understanding all changes in PR.

Comment (DaanHoogland) [reply server/src/main/java/org/cloud/network/router/deployment/RouterDeploymentDefinition.java]: ah, i missed the side effect of `verifyServiceOfferingByUuid()`, where the id is set, thanks.

Maybe to make this more obvious for stupid people like me we can make `verifyServiceOfferingByUuid()` return a uuid or null and set it in the `findAccountServiceOfferingId()`?

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@shwstppr do you 👍 this, based on your tests?

**Review by shwstppr — APPROVED** [review]
LGTM. Tested in a local env.

- [x]  Adding new compute offering and setting it in new global setting results in VR using default offering
- [x] Adding new system service offering for routers and setting it in new global setting results in VR using new offering
- [x] Setting global setting empty results in VR using default offering as expected

Test failures don't seem related. Will be great if we can have corresponding doc PR for user documentation if not there already @weizhouapache

**Comment by weizhouapache** [comment]
> LGTM. Tested in a local env.
> 
> * [x]   Adding new compute offering and setting it in new global setting results in VR using default offering
> * [x]  Adding new system service offering for routers and setting it in new global setting results in VR using new offering
> * [x]  Setting global setting empty results in VR using default offering as expected
> 
> Test failures don't seem related. Will be great if we can have corresponding doc PR for user documentation if not there already @weizhouapache

@shwstppr thanks for testing. Will add doc if this is merged.

**Review by DaanHoogland — APPROVED** [review]

**Comment by DaanHoogland** [comment]
@weizhouapache is this still up to date?

**Comment by weizhouapache** [comment]
> @weizhouapache is this still up to date?

@DaanHoogland yes.

**Comment by DaanHoogland** [comment]
@weizhouapache I see consistent smoke test errors for redundant VPCs since this weekend and this is the only related PR I have found so far. can you have a look e.g. at https://github.com/apache/cloudstack/pull/4200#issuecomment-713227303
@blueorangutan package

**Comment by weizhouapache** [comment]
> @weizhouapache I see consistent smoke test errors for redundant VPCs since this weekend and this is the only related PR I have found so far. can you have a look e.g. at [#4200 (comment)](https://github.com/apache/cloudstack/pull/4200#issuecomment-713227303)
> @blueorangutan package

@DaanHoogland 
if you search "test_01_create_redundant_VPC_2tiers_4VMs_4IPs_4PF_ACL" in your mailbox, you may get many emails :-D
it is not caused by this pr.
as I said in https://github.com/apache/cloudstack/pull/4386#issuecomment-706380544, it is caused by keepalived issue, which should be fixed by #4413

**Comment by DaanHoogland** [comment]
thanks @weizhouapache , i saw it happening a lot suddenly overnight and got a little worried I had merged something overhastely, sorry, never mind.

**Pull request merged**